lib/src/data/repository/api_repository.dart

```diff
@@ -0,0 +1,152 @@
+  static const String apiKeyUri = '?api_key=$apiKey';
+
+  late final String addingUri;
+  static const String latest = '/latest';
+  static const String popular = '/popular';
+  static const String upcoming = '/upcoming';
+  static const String nowPlaying = '/now_playing';
+  static const String cast = '/credits';
+  final Client client;
+
+  late final int movieId;
+
+  APIRepository({required this.client});
+
+  APIRepository.withId({required int movieId, required this.client}) {
+    this.movieId = movieId;
+  }
+
+  @override
+  Future<DataState<List<Genre>>> loadGenres() async {
+    List<GenreModel> genres = <GenreModel>[];
+    try {
+      final Response response = await client.get(
+        Uri.parse(
+          '$genresUri$apiKeyUri',
+        ),
+      );
+      if (response.statusCode == HttpStatus.ok) {
+        final List<Map<String, dynamic>> genreList =
+            List<Map<String, dynamic>>.from(
+          json.decode(response.body)['genres'],
+        );
+        genreList.forEach(
+          (Map<String, dynamic> genreJson) {
+            GenreModel genre = GenreModel.fromJson(genreJson);
+            genres.add(genre);
+          },
+        );
+        return DataSuccess<List<Genre>>(data: genres);
+      } else {
+        return DataFailure<List<Genre>>(
+            error: Exception('Failed to get genre data'),);
+      }
+    } catch (e) {
+      rethrow;
+    }
+  }
+
+  Future<DataState<List<MovieModel>>> _fetchMovies(String url) async {
+    try {
+      final Response response = await client.get(
+        Uri.parse(
+          url,
+        ),
+      );
+      if (response.statusCode == HttpStatus.ok) {
+        return DataSuccess<List<MovieModel>>(
+          data: TMDBApiModel.fromJson(
+            json.decode(response.body),
+          ).results,
+        );
+      } else {
+        return DataFailure<List<MovieModel>>(
+          error: Exception(messageFetchMoviesError),
+        );
+      }
+    } catch (error) {
+      rethrow;
+    }
+  }
+
+  Future<DataState<List<Movie>>> loadMoviesByType(Endpoint endpoint) async {
+    late final String search;
+    switch(endpoint){
+      case Endpoint.popular:
+        search = popular;
+      case Endpoint.nowPlaying:
+        search = nowPlaying;
+      case Endpoint.upcoming:
+        search = upcoming;
+    }
+    return await _fetchMovies('$moviesDetailsUri$search$apiKeyUri');
+  }
+
+  @override
+  Future<DataState<List<Movie>>> loadMovies() async {
+    return await _fetchMovies('$moviesDetailsUri$popular$apiKeyUri');
+  }
+
+  @override
+  Future<DataState<List<Movie>>> loadPopularMovies() {
+    return _fetchMovies('$moviesDetailsUri$popular$apiKeyUri');
+  }
+
+  @override
+  Future<DataState<List<Movie>>> loadNowPlayingMovies() {
+    return _fetchMovies('$moviesDetailsUri$nowPlaying$apiKeyUri');
+  }
+
+  @override
+  Future<DataState<List<Movie>>> loadUpcomingMovies() {
+    return _fetchMovies('$moviesDetailsUri$upcoming$apiKeyUri');
+  }
```

Comment:
You don't need to have the same method to every method call that you have, you can use directly use the uri as a parameter

Reply:
I already have it but i guess it's not okay yo have the two things, so i deleted this implementation

Reply:
Now i have this thing that in the use case i have to call the method `loadMoviesByType(Endpoint.anything)`. It's that okay right?

Reply:
Yes, it's correct but be carefull with the names